pubtools/_quay/image_untagger.py

```diff
@@ -0,0 +1,213 @@
+            else:
+                image = image_schema.format(
+                    self.host, repository, attributes["manifest_digest"]
+                )
+                manifest_list = self._quay_client.get_manifest_list(image)
+                tag_digest_mapping[tag] = [attributes["manifest_digest"]]
+                digest_tag_mapping.setdefault(attributes["manifest_digest"], []).append(
+                    tag
+                )
+
+                for manifest in manifest_list["manifests"]:
+                    tag_digest_mapping[tag].append(manifest["digest"])
+                    digest_tag_mapping.setdefault(manifest["digest"], []).append(tag)
+
+        return (tag_digest_mapping, digest_tag_mapping)
+
+    def get_lost_digests(self, tags, tag_digest_mapping, digest_tag_mapping):
+        """
+        Calculate a list of digests that would be lost if the provided tags were removed.
+
+        Args:
+            tags ([str]):
+                List of tags that would be removed from a repo.
+            tag_digest_mapping ({str: [str]}):
+                Mapping of which digests are referenced by a given tag.
+            digest_tag_mapping ({str: [str]}):
+                Mapping of which tags reference a given digest.
+
+        Returns ([str]):
+            Digests that would be lost if given tags were removed.
+        """
+        remove_digests = []
+        lost_digests = []
+        for tag in tags:
+            for digest in tag_digest_mapping.get(tag, []):
+                if digest not in remove_digests:
+                    remove_digests.append(digest)
+
+        for digest in remove_digests:
+            # which tags would remain referencing a given digest?
+            remaining_tags = set(digest_tag_mapping[digest]) - set(tags)
+            if len(remaining_tags) == 0:
+                lost_digests.append(digest)
+
+        return lost_digests
+
+    def untag_images(self):
+        """
+        Determine if the specified tags may be removed and remove them.
+
+        Returns ([str]):
+            List of image references lost by the untagging.
+        """
+        if not self._quay_client:
+            raise RuntimeError("QuayClient instance must be set")
+
+        repo_tag_mapping = self.get_repository_tags_mapping()
+        schema = "{0}/{1}@{2}"
+        lost_imgs = []
+
+        for repo in repo_tag_mapping:
+            tag_digest_mapping, digest_tag_mapping = self.construct_tag_digest_mappings(
+                repo
+            )
+
+            lost_digests = self.get_lost_digests(
+                repo_tag_mapping[repo], tag_digest_mapping, digest_tag_mapping
+            )
+            if len(lost_digests) > 0:
+                lost_imgs += [
+                    schema.format(self.host, repo, digest) for digest in lost_digests
+                ]
+
+        if not self.remove_last and lost_imgs:
+            raise ValueError(
+                "Following images would no longer be referencable by tag after"
```

Comment:
This sounds like images may still be reachable by digests. Is it worth also mentioning "images will be deleted when no tag references them"?

Reply:
I'm not really sure if the images will be actually deleted. If I uploaded a manifest list containing their digest to some tag, I think they would become referencable again. Although there's likely some garbage-collection in place.

Reply:
It's configurable at the namespace level. Generally will be e.g. 14 days.